File: algorithms/sorting/merge_sort.py
# Merge sort algorithm implementation
def merge(arr, left, right):
    left_i = right_i = arr_i = 0
    
    while left_i < len(left) and right_i < len(right):
        if left[left_i] < right[right_i]:
            arr[arr_i] = left[left_i]
            left_i += 1
        else:
            arr[arr_i] = right[right_i]
            right_i += 1
        arr_i += 1

    while left_i < len(left):
        arr[arr_i] = left[left_i]
        left_i += 1
        arr_i += 1

    while right_i < len(right):
        arr[arr_i] = right[right_i]
        right_i += 1
        arr_i += 1

def merge_sort(arr):
    print("arr: ", arr)
    if len(arr) == 1:
        return arr
    
    mid = len(arr) // 2
    left = arr[:mid]
    right = arr[mid:]

    merge_sort(left)
    merge_sort(right)
    merge(arr, left, right)

    return arr 

# [12, 11, 13, 5, 6, 7, 1, 9]
# l = [12, 11, 13, 5]
# r = [6, 7, 1, 9]

# [12, 11, 13, 5]
# l = [12, 11]
# r = [13, 5]

# [12]
# [11]
# [11, 12]

# [13, 5]
# [13]
# [5]
# [5, 13]

# [5, 11, 12, 13]

if __name__ == '__main__':
    arr = [12, 11, 13, 5, 6, 7, 1, 9, 1]
    print("Given array is", arr)
    merge_sort(arr)
    print("Sorted array is: ", arr)

File: algorithms/sorting/merge_sort.cpp
// Merge Sort C++ implementation

#include <iostream>
#include <vector>

using namespace std;

void merge(vector<int>& arr, int l, int m, int r) {
    /**
     * arr: the array where the merged elements will be stored
     * l: left index
     * m: middle index
     * r: right index
     */
    int n1 = m - l + 1;
    int n2 = r - m;

    vector<int> L(n1), R(n2);

    for (int i = 0; i < n1; i++) {
        L[i] = arr[l + i];
    }

    for (int i = 0; i < n2; i++) {
        R[i] = arr[m + 1 + i];
    }

    int i = 0, j = 0, k = l;

    while (i < n1 && j < n2) {
        if (L[i] <= R[j]) {
            arr[k++] = L[i++];
        } else {
            arr[k++] = R[j++];
        }
    }

    while (i < n1) {
        arr[k++] = L[i++];
    }

    while (j < n2) {
        arr[k++] = R[j++];
    }
}

void mergeSort(vector<int>& arr, int l, int r) {
    if (l < r) {
        int m = l + (r - l) / 2;

        mergeSort(arr, l, m);
        mergeSort(arr, m + 1, r);

        merge(arr, l, m, r);
    }
}

int main() {
    vector<int> arr = {12, 11, 13, 5, 6, 7};

    mergeSort(arr, 0, arr.size() - 1);

    for (int i = 0; i < arr.size(); i++) {
        cout << arr[i] << " ";
    }

    return 0;
}
